### PR TITLE
LP-2840 integrate team member invite

### PR DESCRIPTION
**Pull request by annas-usman**
[LP-2840](https://philanthropyu.atlassian.net/browse/LP-2840)

[Theme PR](https://github.com/OmnipreneurshipAcademy/adg-edx-theme/pull/508)

***

#### Description:
To enable the functionality of inviting team members helpers.py has been created along with a form for Invitation and a common post method for both kind of views.

**Note:** The tests will be added in an upcoming ticket

### CODE REVIEW

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2876/workflows/workflows_id/jobs/2876/artifacts)**. Coverage !!! :scream:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/admin_settings/constants.py (100%)
openedx/adg/admin_dashboard/admin_settings/forms.py (100%)
openedx/adg/admin_dashboard/admin_settings/helpers.py (47.9%): Missing lines 113,115,117-118,120-121,123,128,143-144,146-147,149-150,152,154-155,157,172-173,175,177-179,182,184,197,199-200,204-205,207,220,222-224,226-229,231,233,247-250,253,264,267
openedx/adg/admin_dashboard/admin_settings/tests/test_views.py (100%)
openedx/adg/admin_dashboard/admin_settings/views.py (68.2%): Missing lines 36-41,43
openedx/adg/common/lib/mandrill_client/client.py (100%)
-------------
Total:   132 lines
Missing: 56 lines
Coverage: 57%
-------------
```

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2882/workflows/workflows_id/jobs/2882/artifacts)**. Coverage !!! :scream:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/admin_settings/constants.py (100%)
openedx/adg/admin_dashboard/admin_settings/forms.py (100%)
openedx/adg/admin_dashboard/admin_settings/helpers.py (56.2%): Missing lines 25-28,31,42,45
openedx/adg/admin_dashboard/admin_settings/models.py (100%)
openedx/adg/admin_dashboard/admin_settings/tests/test_views.py (100%)
openedx/adg/admin_dashboard/admin_settings/views.py (42.5%): Missing lines 51-56,58,71,73,75-76,78-79,81,86,100-101,103-104,106-107,109,111-112,114,128-129,131-132,134-136,139,141,153,155-156,160-161,163,175,177-179,181-184,186,188
openedx/adg/admin_dashboard/applications/mixins.py (66.7%): Missing lines 54
openedx/adg/common/lib/mandrill_client/client.py (100%)
openedx/adg/lms/applications/models.py (100%)
-------------
Total:   134 lines
Missing: 58 lines
Coverage: 56%
-------------
```

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 1]
```diff
@@ -1,31 +1,226 @@
...
+    def post(self, request, *args, **kwargs):
+        """
+        Post method to parse, validate and send invitations to the invitees
+        """
+        data = request.POST.dict()
+        additional_context = self.handle_invitations(data)
+        queryset = self.get_queryset()
+        context = self.get_context_data(object_list=queryset)
+        context['invitation_error'] = additional_context['invitation_error']
+        context['invitations'] = additional_context['invitations']
```
Comment (javeriaejaz-arbisoft) [inline openedx/adg/admin_dashboard/admin_settings/views.py:56]: Can we please handle context in separate context function?

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 2]
```diff
@@ -1,31 +1,226 @@
...
+        for key, value in invitations_data.items():
+            field, invitation_id = key.split('-')
+
+            if invitation_id in invitations:
+                invitations[invitation_id].update({field: value})
+            else:
+                invitations[invitation_id] = {field: value}
+
+            if field == EMAIL:
+                email_validation_error = self.validate_admin_email(value, email_validation_error)
```
Comment (javeriaejaz-arbisoft) [inline openedx/adg/admin_dashboard/admin_settings/views.py:112]: Why are we validating emails here and not in form?

Comment (javeriaejaz-arbisoft) [reply openedx/adg/admin_dashboard/admin_settings/views.py:112]: Moving validations in form will make the view a lot cleaner.

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 3]
```diff
@@ -1,31 +1,226 @@
...
+            statuses = [Invitation.PENDING, Invitation.ACCEPTED]
+
+            if Invitation.objects.filter(email=email, status__in=statuses).exists():
+                email_validation_error = ALREADY_INVITED_ADMIN_ERROR.format(email=email)
+            elif user and (
+                is_superuser_or_adg_admin(user) or BusinessLine.is_user_business_line_admin(user)
+            ):
+                email_validation_error = ALREADY_ADMIN_ERROR.format(email=email)
+
+        return email_validation_error
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/views.py:141]: How about?
```suggestion
        email_validation_error = INVALID_EMAILS_ERROR if get_email_validation_error(email) else ''

        if email_validation_error:
            return email_validation_error

        user = User.objects.filter(email=email).first()
        statuses = (Invitation.PENDING, Invitation.ACCEPTED)

        if Invitation.objects.filter(email=email, status__in=statuses).exists():
            email_validation_error = ALREADY_INVITED_ADMIN_ERROR.format(email=email)
        elif user and has_admin_permissions(user):
            email_validation_error = ALREADY_ADMIN_ERROR.format(email=email)

        return email_validation_error
```

**Review by asadali145 — CHANGES_REQUESTED** [review]

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2883/workflows/workflows_id/jobs/2883/artifacts)**. Coverage !!! :scream:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/admin_settings/constants.py (100%)
openedx/adg/admin_dashboard/admin_settings/forms.py (46.4%): Missing lines 33-35,42-44,46-47,50-51,59-61,63-64
openedx/adg/admin_dashboard/admin_settings/helpers.py (38.7%): Missing lines 33-35,37-39,41-43,56-58,72-75,78,89,92
openedx/adg/admin_dashboard/admin_settings/models.py (100%)
openedx/adg/admin_dashboard/admin_settings/tests/test_views.py (100%)
openedx/adg/admin_dashboard/admin_settings/views.py (50.7%): Missing lines 44-46,59-61,63-67,81,83-84,86-87,89,91,103-104,106-109,111-113,115-119,121
openedx/adg/admin_dashboard/applications/mixins.py (66.7%): Missing lines 54
openedx/adg/common/lib/mandrill_client/client.py (100%)
openedx/adg/lms/applications/models.py (100%)
-------------
Total:   150 lines
Missing: 68 lines
Coverage: 54%
-------------
```

File: openedx/adg/admin_dashboard/admin_settings/constants.py [hunk 4]
```diff
@@ -6,3 +6,14 @@
 ADMIN_SETTINGS_PAGE_SIZE = 10
 ADG_ADMIN_DISPLAY_STRING = _('ADG Admin')
 BU_ADMIN_DISPLAY_STRING = _('BU Admin')
+
+INVALID_EMAILS_ERROR = _('Please enter valid emails.')
+ALREADY_INVITED_ADMIN_ERROR = _('{email} has already been sent an invitation.')
+ALREADY_ADMIN_ERROR = _('{email} is already an admin.')
+PERMISSION_REQUIRED_ERROR = _('Please select Permissions.')
+BUSINESS_UNIT_REQUIRED_ERROR = _('Please select Business Units')
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/constants.py:14]: Add the full stop at the end.

File: openedx/adg/admin_dashboard/admin_settings/forms.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        fields = ['email', 'business_unit', 'permission', 'status']
+
+    def clean(self):
+        """
+        Raise error if there's already pending or accepted invitation against this email or a user is registered using
+        this email as an admin.
+        """
+        cleaned_data = super().clean()
+        self.validate_email(cleaned_data)
+        self.validate_business_unit(cleaned_data)
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/forms.py:35]: We should not change the return type of a method that we are overriding, this should return cleaned_data at the end.

File: openedx/adg/admin_dashboard/admin_settings/forms.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        Raise error if user selected BU Admin but did not select any Business Unit.
+        """
+        if PERMISSION in cleaned_data:
+            permission = cleaned_data.get(PERMISSION)
+            business_unit = cleaned_data.get(BUSINESS_UNIT)
+
+            if permission == Invitation.BU_ADMIN and business_unit is None:
+                raise forms.ValidationError(
+                    {BUSINESS_UNIT: BUSINESS_UNIT_REQUIRED_ERROR}
+                )
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/forms.py:66]:
```suggestion
        if PERMISSION not in cleaned_data:
            return 

        permission = cleaned_data.get(PERMISSION)
        business_unit = cleaned_data.get(BUSINESS_UNIT)

        if permission == Invitation.BU_ADMIN and business_unit is None:
            raise forms.ValidationError(
                {BUSINESS_UNIT: BUSINESS_UNIT_REQUIRED_ERROR}
            )
```

File: openedx/adg/admin_dashboard/admin_settings/helpers.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+         invitation (dict): dictionary with information related to a single invitation
+
+    Returns:
+        str: registration url with added query parameter
+    """
+    root_url = configuration_helpers.get_value('LMS_ROOT_URL', settings.LMS_ROOT_URL)
+    registration_url = f'{root_url}{reverse("register_user")}'
+    business_unit = invitation[BUSINESS_UNIT] if BUSINESS_UNIT in invitation else None
+    query_parameters = (f'?email={invitation[EMAIL]}&permission={invitation[PERMISSION]}&'
+                        f'business_unit={business_unit}')
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/helpers.py:76]:
```suggestion
    query_parameters = (
        f'?email={invitation[EMAIL]}&permission={invitation[PERMISSION]}&business_unit={business_unit}'
    )
```

File: openedx/adg/admin_dashboard/admin_settings/helpers.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    Arguments:
+        html (ErrorList): form error of a field
+
+    Returns:
+        str: error
+    """
+    error = re.findall(r'<li>.+<\/li>', str(html))[0]
+    error = error.replace('<li>', '').replace('</li>', '')
+    return error
+
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/helpers.py:59]: As discussed we can use form.errors.as_data().

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 9]
```diff
@@ -1,31 +1,159 @@
...
+
+        for key, value in invitations_data.items():
+            field, invitation_id = key.split('-')
+
+            if invitation_id in invitations:
+                invitations[invitation_id].update({field: value})
+            else:
+                invitations[invitation_id] = {field: value}
+
+        return invitations
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/views.py:91]: We can use collections.defaultdict here.

Code will be:
```suggestion
        invitations = defaultdict(dict)

        for key, value in invitations_data.items():
            field, invitation_id = key.split('-')
            invitations[invitation_id].update({field: value})

        return invitations
```

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 10]
```diff
@@ -1,31 +1,159 @@
...
+
+    def validate_and_save_invitations(self, invitations):
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/views.py:93]: Pls follow SRP, This should only validate and some other method should save.

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 11]
```diff
@@ -1,31 +1,159 @@
...
+                invalid_form_error = extract_form_error(form.errors)
+                break
+
+        if not invalid_form_error:
+            for form, invitation in zip(forms, invitations.values()):
+                form.save()
+                registration_url = create_registration_url(invitation)
+                send_team_member_invitation_email(invitation[EMAIL], registration_url)
+
+        return invalid_form_error
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/views.py:121]:
```suggestion
        forms = []

        for invitation in invitations.values():
            invitation[STATUS] = Invitation.PENDING
            form = InvitationForm(invitation)

            if not form.is_valid():
                invalid_form_error = extract_form_error(form.errors)
                return invalid_form_error

            forms.append(form)

        for form, invitation in zip(forms, invitations.values()):
            form.save()
            registration_url = create_registration_url(invitation)
            send_team_member_invitation_email(invitation[EMAIL], registration_url)
```

File: openedx/adg/admin_dashboard/admin_settings/views.py [hunk 12]
```diff
@@ -1,31 +1,159 @@
...
+            user (User): admin for which the context is to be updated
+
+        Returns:
+            dict: updated context dictionary
+        """
+        business_lines = BusinessLine.business_lines_for_user(user)
+        business_line_titles = {}
+
+        for business_line in business_lines:
+            business_line_titles[business_line.id] = business_line.title
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/admin_settings/views.py:138]:
```suggestion
        business_line_titles = {
            business_line.id: business_line.title
            for business_line in business_lines
        }
```

**Review by asadali145 — COMMENTED** [review]
Great effort to get this revamped for validations in a night, Here are a few code improvements which will make our code more cleaner. Its already quite clean though :)

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2884/workflows/workflows_id/jobs/2884/artifacts)**. Coverage !!! :scream:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/admin_settings/constants.py (100%)
openedx/adg/admin_dashboard/admin_settings/forms.py (43.3%): Missing lines 33-36,43-45,47-48,51-52,60-61,63-64,66-67
openedx/adg/admin_dashboard/admin_settings/helpers.py (40.0%): Missing lines 31-33,35-37,39-40,54-57,61,72,75
openedx/adg/admin_dashboard/admin_settings/models.py (100%)
openedx/adg/admin_dashboard/admin_settings/tests/test_views.py (100%)
openedx/adg/admin_dashboard/admin_settings/views.py (52.3%): Missing lines 45-47,60-62,64-68,82,84-86,88,101-102,104-106,108-110,112,114,116,127-130
openedx/adg/admin_dashboard/applications/mixins.py (66.7%): Missing lines 54
openedx/adg/common/lib/mandrill_client/client.py (100%)
openedx/adg/lms/applications/models.py (100%)
-------------
Total:   144 lines
Missing: 64 lines
Coverage: 55%
-------------
```

File: openedx/adg/admin_dashboard/admin_settings/forms.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    def validate_email(self, cleaned_data):
+        """
+        Raise error if there's already pending or accepted invitation against this email or a user is registered using
+        this email as an admin.
+        """
+        email = cleaned_data.get('email')
+        user = User.objects.filter(email=email).first()
+        statuses = (Invitation.PENDING, Invitation.ACCEPTED)
+
+        if Invitation.objects.filter(email=email, status__in=statuses).exists():  # pylint: disable=no-else-raise
```
Comment (javeriaejaz-arbisoft) [inline openedx/adg/admin_dashboard/admin_settings/forms.py:47]: By writing two separate if you can avoid pylint warning.

File: openedx/adg/admin_dashboard/admin_settings/views.py (hunk 10 above)
Comment (javeriaejaz-arbisoft) [reply openedx/adg/admin_dashboard/admin_settings/views.py:90]: You can directly call save method in handle_invitations.

**Review by javeriaejaz-arbisoft — APPROVED** [review]
Nicely done. Added only few suggestions and great effort in revamping it overnight 👏🏼.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2887/workflows/workflows_id/jobs/2887/artifacts)**. Coverage !!! :scream:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/admin_settings/constants.py (100%)
openedx/adg/admin_dashboard/admin_settings/forms.py (43.3%): Missing lines 32-35,42-44,46-47,51-52,60-61,63-64,66-67
openedx/adg/admin_dashboard/admin_settings/helpers.py (40.0%): Missing lines 31-33,35-37,39-40,54-57,61,72,75
openedx/adg/admin_dashboard/admin_settings/models.py (100%)
openedx/adg/admin_dashboard/admin_settings/tests/test_views.py (100%)
openedx/adg/admin_dashboard/admin_settings/views.py (51.5%): Missing lines 45-47,60-62,64-65,67-71,85,87-89,91,104-105,107-109,111-113,115,117,128-131
openedx/adg/admin_dashboard/applications/mixins.py (66.7%): Missing lines 54
openedx/adg/common/lib/mandrill_client/client.py (100%)
openedx/adg/lms/applications/models.py (100%)
-------------
Total:   145 lines
Missing: 65 lines
Coverage: 55%
-------------
```

**Review by asadali145 — APPROVED** [review]
LGTM! Great Effort!!!

**Pull request merged**